File: src/TodoTracker.Core/Pomodoro.cs
namespace TodoTracker.Core;

public enum PomodoroPhase
{
    Idle,
    Focus,
    ShortBreak,
    LongBreak,
}

public enum PomodoroEventKind
{
    FocusCompleted,
    BreakCompleted,
}

public sealed record PomodoroEvent(PomodoroEventKind Kind, Guid? ItemId, DateTimeOffset At);

public sealed record PomodoroSettings
{
    public PomodoroSettings(int focusMinutes = 25, int shortBreakMinutes = 5, int longBreakMinutes = 15, int focusesBeforeLongBreak = 4)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(focusMinutes);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(shortBreakMinutes);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(longBreakMinutes);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(focusesBeforeLongBreak);
        FocusMinutes = focusMinutes;
        ShortBreakMinutes = shortBreakMinutes;
        LongBreakMinutes = longBreakMinutes;
        FocusesBeforeLongBreak = focusesBeforeLongBreak;
    }

    public int FocusMinutes { get; }

    public int ShortBreakMinutes { get; }

    public int LongBreakMinutes { get; }

    public int FocusesBeforeLongBreak { get; }

    public TimeSpan DurationOf(PomodoroPhase phase) => phase switch
    {
        PomodoroPhase.ShortBreak => TimeSpan.FromMinutes(ShortBreakMinutes),
        PomodoroPhase.LongBreak => TimeSpan.FromMinutes(LongBreakMinutes),
        _ => TimeSpan.FromMinutes(FocusMinutes),
    };
}

/// <summary>
/// Pomodoro state machine. Breaks start automatically after a focus session; a new focus session
/// always needs an explicit start so the user stays in control.
/// </summary>
public sealed class PomodoroTimer
{
    public PomodoroTimer(PomodoroSettings? settings = null)
    {
        Settings = settings ?? new PomodoroSettings();
    }

    public PomodoroSettings Settings { get; internal set; }

    public PomodoroPhase Phase { get; private set; } = PomodoroPhase.Idle;

    public DateTimeOffset? EndsAt { get; private set; }

    public TimeSpan? PausedRemaining { get; private set; }

    public Guid? ItemId { get; private set; }

    public int CompletedFocusCount { get; private set; }

    public bool IsRunning => EndsAt is not null;

    public TimeSpan Remaining(DateTimeOffset now)
    {
        if (EndsAt is { } endsAt)
        {
            var left = endsAt - now;
            return left > TimeSpan.Zero ? left : TimeSpan.Zero;
        }

        return PausedRemaining ?? Settings.DurationOf(Phase);
    }

    public void StartFocus(DateTimeOffset now, Guid? itemId = null)
    {
        Phase = PomodoroPhase.Focus;
        ItemId = itemId;
        PausedRemaining = null;
        EndsAt = now + Settings.DurationOf(PomodoroPhase.Focus);
    }

    public void Pause(DateTimeOffset now)
    {
        if (EndsAt is null || Phase == PomodoroPhase.Idle)
        {
            return;
        }

        PausedRemaining = Remaining(now);
        EndsAt = null;
    }

    public void Resume(DateTimeOffset now)
    {
        if (PausedRemaining is not { } remaining || Phase == PomodoroPhase.Idle)
        {
            return;
        }

        EndsAt = now + remaining;
        PausedRemaining = null;
    }

    public void Reset()
    {
        Phase = PomodoroPhase.Idle;
        EndsAt = null;
        PausedRemaining = null;
        ItemId = null;
        CompletedFocusCount = 0;
    }

    /// <summary>Skips the current phase: focus goes to a (short) break without counting; a break ends.</summary>
    public void Skip(DateTimeOffset now)
    {
        if (Phase == PomodoroPhase.Focus)
        {
            BeginBreak(PomodoroPhase.ShortBreak, now);
        }
        else if (Phase != PomodoroPhase.Idle)
        {
            GoIdle();
        }
    }

    /// <summary>Advances at most one phase transition. Call repeatedly to catch up after sleep.</summary>
    public PomodoroEvent? Tick(DateTimeOffset now)
    {
        if (EndsAt is not { } endsAt || now < endsAt)
        {
            return null;
        }

        if (Phase == PomodoroPhase.Focus)
        {
            CompletedFocusCount++;
            var item = ItemId;
            var breakPhase = CompletedFocusCount % Settings.FocusesBeforeLongBreak == 0 ? PomodoroPhase.LongBreak : PomodoroPhase.ShortBreak;
            BeginBreak(breakPhase, endsAt);
            return new PomodoroEvent(PomodoroEventKind.FocusCompleted, item, endsAt);
        }

        var itemId = ItemId;
        GoIdle();
        return new PomodoroEvent(PomodoroEventKind.BreakCompleted, itemId, endsAt);
    }

    internal void DetachItem() => ItemId = null;

    internal void Restore(PomodoroPhase phase, DateTimeOffset? endsAt, TimeSpan? pausedRemaining, Guid? itemId, int completedFocusCount)
    {
        Phase = phase;
        EndsAt = phase == PomodoroPhase.Idle ? null : endsAt;
        PausedRemaining = phase == PomodoroPhase.Idle || endsAt is not null ? null : pausedRemaining;
        ItemId = itemId;
        CompletedFocusCount = Math.Max(0, completedFocusCount);
    }

    private void BeginBreak(PomodoroPhase phase, DateTimeOffset from)
    {
        Phase = phase;
        PausedRemaining = null;
        EndsAt = from + Settings.DurationOf(phase);
    }

    private void GoIdle()
    {
        Phase = PomodoroPhase.Idle;
        EndsAt = null;
        PausedRemaining = null;
    }
}

File: apple/TodoTrackerKit/Sources/TodoTrackerKit/Pomodoro.swift
import Foundation

public enum PomodoroPhase: String, Codable, Sendable {
    case idle, focus, shortBreak, longBreak
}

public enum PomodoroEventKind: String, Sendable {
    case focusCompleted, breakCompleted
}

public struct PomodoroEvent: Equatable, Sendable {
    public let kind: PomodoroEventKind
    public let itemId: UUID?
    public let at: Date
}

public struct PomodoroSettings: Codable, Equatable, Sendable {
    public var focusMinutes = 25
    public var shortBreakMinutes = 5
    public var longBreakMinutes = 15
    public var focusesBeforeLongBreak = 4

    public init(focusMinutes: Int = 25, shortBreakMinutes: Int = 5, longBreakMinutes: Int = 15, focusesBeforeLongBreak: Int = 4) {
        self.focusMinutes = max(1, focusMinutes)
        self.shortBreakMinutes = max(1, shortBreakMinutes)
        self.longBreakMinutes = max(1, longBreakMinutes)
        self.focusesBeforeLongBreak = max(1, focusesBeforeLongBreak)
    }

    public func duration(of phase: PomodoroPhase) -> TimeInterval {
        switch phase {
        case .shortBreak: return TimeInterval(shortBreakMinutes * 60)
        case .longBreak: return TimeInterval(longBreakMinutes * 60)
        default: return TimeInterval(focusMinutes * 60)
        }
    }
}

/// Pomodoro state machine. Breaks start automatically after focus; a new focus block needs an explicit start.
public struct PomodoroTimer: Equatable, Sendable {
    public var settings = PomodoroSettings()
    public private(set) var phase: PomodoroPhase = .idle
    public private(set) var endsAt: Date?
    public private(set) var pausedRemaining: TimeInterval?
    public private(set) var itemId: UUID?
    public private(set) var completedFocusCount = 0

    public init(settings: PomodoroSettings = PomodoroSettings()) {
        self.settings = settings
    }

    public var isRunning: Bool { endsAt != nil }

    public func remaining(_ now: Date) -> TimeInterval {
        if let endsAt { return max(0, endsAt.timeIntervalSince(now)) }
        return pausedRemaining ?? settings.duration(of: phase)
    }

    public mutating func startFocus(now: Date, itemId: UUID? = nil) {
        phase = .focus
        self.itemId = itemId
        pausedRemaining = nil
        endsAt = now.addingTimeInterval(settings.duration(of: .focus))
    }

    public mutating func pause(now: Date) {
        guard endsAt != nil, phase != .idle else { return }
        pausedRemaining = remaining(now)
        endsAt = nil
    }

    public mutating func resume(now: Date) {
        guard let left = pausedRemaining, phase != .idle else { return }
        endsAt = now.addingTimeInterval(left)
        pausedRemaining = nil
    }

    public mutating func reset() {
        phase = .idle
        endsAt = nil
        pausedRemaining = nil
        itemId = nil
        completedFocusCount = 0
    }

    public mutating func skip(now: Date) {
        if phase == .focus {
            beginBreak(.shortBreak, from: now)
        } else if phase != .idle {
            goIdle()
        }
    }

    /// Advances at most one transition; call repeatedly to catch up after sleep.
    public mutating func tick(now: Date) -> PomodoroEvent? {
        guard let end = endsAt, now >= end else { return nil }
        if phase == .focus {
            completedFocusCount += 1
            let item = itemId
            beginBreak(completedFocusCount % settings.focusesBeforeLongBreak == 0 ? .longBreak : .shortBreak, from: end)
            return PomodoroEvent(kind: .focusCompleted, itemId: item, at: end)
        }
        let item = itemId
        goIdle()
        return PomodoroEvent(kind: .breakCompleted, itemId: item, at: end)
    }

    mutating func detachItem() { itemId = nil }

    mutating func restore(phase: PomodoroPhase, endsAt: Date?, pausedRemaining: TimeInterval?, itemId: UUID?, completedFocusCount: Int) {
        self.phase = phase
        self.endsAt = phase == .idle ? nil : endsAt
        self.pausedRemaining = (phase == .idle || endsAt != nil) ? nil : pausedRemaining
        self.itemId = itemId
        self.completedFocusCount = max(0, completedFocusCount)
    }

    private mutating func beginBreak(_ phase: PomodoroPhase, from: Date) {
        self.phase = phase
        pausedRemaining = nil
        endsAt = from.addingTimeInterval(settings.duration(of: phase))
    }

    private mutating func goIdle() {
        phase = .idle
        endsAt = nil
        pausedRemaining = nil
    }
}
